### PR TITLE
Drivers: clock: EPIT: Add driver for the EPIT

### PR DESCRIPTION
**Pull request by puranjaymohan**
Add the driver for EPIT timer/counter for the I.MX boards.
This solves #1767 
Signed-off-by: Puranjay Mohan <puranjay12@gmail.com>

### CODE REVIEW

File: src/drivers/clock/epit/epit.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+#define EPIT_CMPR   	(EPIT_BASE + 0x0C) /* Compare Register */
+#define EPIT_CNR    	(EPIT_BASE + 0x10) /* Counter Register */
+
+#define EPIT_CR_PRESCLR (0x42 << 4)/* Prescaler value of 66 */
+#define EPIT_CR_EN 		(1 << 0) /* Enable Bit */
+#define EPIT_CR_ENMOD   (1 << 1) /* Enable Mode Bit */
+#define EPIT_CR_OCIEN   (1 << 2) /* Output compare Interrupt Enable bit*/
+#define EPIT_CR_RLD		(1 << 3) /* Counter Reload Control bit */
+#define EPIT_SR_CLEAR	(1 << 0) /* Clear the status register*/
+#define EPIT_CR_CLKSRC	(1 << 24) /* Use peripheral clock as input, 66MHz */
```
Comment (alexkalmuk) [inline src/drivers/clock/epit/epit.c]: Can you please fix formatting?

File: src/drivers/clock/epit/epit.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) & (~EPIT_CR_EN)); /* disable the timer*/
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_CLKSRC); /* Select the clock source */
+	REG32_STORE(EPIT_SR, EPIT_SR_CLEAR); /* Clear the status register */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_ENMOD); /* Set enable mode = 1 */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_EN); /* enable the timer */  
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_OCIEN); /* Enable Interrupts */
+
+	/* setting the load and compare registers */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_RLD | EPIT_CR_IOVW | EPIT_CR_PRESCLR); /* Set reload and overrite */
+	REG32_STORE(EPIT_LR, 0x3e8); /* Load regiter set to 1000 */
```
Comment (alexkalmuk) [inline src/drivers/clock/epit/epit.c]: Should it depend on EPIT_TARGET_HZ value?

Comment (alexkalmuk) [reply src/drivers/clock/epit/epit.c]: We can set EPIT_TARGET_HZ to 2000, but this line will be unchanged.

File: src/drivers/clock/epit/epit.c [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	.event_device   = &epit_event,
+	.counter_device = &epit_counter,
+	.read           = clock_source_read,
+};
+
+EMBOX_UNIT_INIT(epit_init);
+
+STATIC_IRQ_ATTACH(EPIT_IRQ, clock_handler, &imx6_clock_source);
+
+PERIPH_MEMORY_DEFINE(epit_mem, EPIT_BASE, 0x10);
```
Comment (alexkalmuk) [inline src/drivers/clock/epit/epit.c]: Are you sure `0x10`? How many registers EPIT has? I see you provide 5 registers, but even 5 regs have `5 * 4 = 20 = 0x14` in size.

File: src/drivers/clock/epit/epit.c (hunk 1 above)
Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: I will do it, my text editor changes it all the time.

File: src/drivers/clock/epit/epit.c (hunk 2 above)
Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: Yes, I hard-coded it just for testing and getting review. I will change it, load value will depend on EPIT_TARGET_HZ and EPIT_CR_PRESCLR.

File: src/drivers/clock/epit/epit.c (hunk 3 above)
Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: I am sorry for this, the EPIT does have 5 registers, I will change It to 0x14.

File: src/drivers/clock/epit/epit.c [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#define EPIT_PRESCALER	66 /* prescaler value of 66 */
+#define EPIT_CR_PRESCLR	(EPIT_PRESCALER << 4) /* Prescaler value of 66 */
+#define EPIT_CR_EN		(1 << 0) /* Enable Bit */
+#define EPIT_CR_ENMOD		(1 << 1) /* Enable Mode Bit */
+#define EPIT_CR_OCIEN		(1 << 2) /* Output compare Interrupt Enable bit*/
+#define EPIT_CR_RLD		(1 << 3) /* Counter Reload Control bit */
+#define EPIT_SR_CLEAR		(1 << 0) /* Clear the status register*/
+#define EPIT_CR_CLKSRC	(1 << 24) /* Use peripheral clock as input, 66MHz */
+#define EPIT_CR_IOVW		(1 << 17) /* Load Immidiate Override*/
+#define EPIT_LR_VALUE		(66000000)/(EPIT_PRESCALER * EPIT_TARGET_HZ)
```
Comment (alexkalmuk) [inline src/drivers/clock/epit/epit.c]: What is `66000000`?

Comment (alexkalmuk) [reply src/drivers/clock/epit/epit.c]: btw, it's safer to write `#define EPIT_LR_VALUE	 (66000000 / (EPIT_PRESCALER * EPIT_TARGET_HZ))`

Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: @alexkalmuk The 66000000 is the frequency of the peripheral clock i.e. 66 Mhz, it is used in the EPIT timer.

Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: I have changed the #define as you specified.
Thanks

Comment (alexkalmuk) [reply src/drivers/clock/epit/epit.c]: Can we use `#define EPIT_FREQ 66000000` then?

**Comment by alexkalmuk** [comment]
I checked EPIT and it works fine!

File: src/drivers/clock/epit/epit.c (hunk 4 above)
Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: Yes, I will change it.

Comment (puranjaymohan) [reply src/drivers/clock/epit/epit.c]: done.

**Comment by puranjaymohan** [comment]
I have made all the final changes, please merge it.
Thanks. :smile:

**Comment by anton-bondarev** [comment]
@puranjaymohan Thanks. Great work!

**Pull request merged**